alacritty/src/event.rs

```diff
@@ -411,8 +413,32 @@ impl<N: Notify + OnResize> Processor<N> {
                 GlutinEvent::RedrawEventsCleared => {
                     *control_flow = ControlFlow::Wait;
 
-                    if event_queue.is_empty() {
-                        return;
+                    let event_queue_empty = event_queue.is_empty();
+
+                    #[cfg(not(any(target_os = "macos", windows)))]
+                    {
+                        let wayland_queue_empty = if event_loop.is_wayland() {
+                            let num_dispatched = self
+                                .display
+                                .wayland_event_queue
+                                .as_mut()
+                                .unwrap()
+                                .dispatch_pending(&mut (), |_, _, _| {})
+                                .unwrap();
+                            num_dispatched == 0
+                        } else {
+                            true
+                        };
+
+                        if event_queue_empty && wayland_queue_empty {
```

Comment:
Why can't we skip processing if the wayland queue is not empty? We don't need to render on every frame callback, right? So shouldn't matter?

Reply:
This is intentional, since without it we were missing redraw in some edge cases. IIRC it could be true for obscure windows and something like that, since compositor could start sending you events to redraw before you're going to switch into that window. This code exists because of a bug of missing redraws in some corner cases, that's all.

Reply:
Also, one thing why it's required. You're processing all your events on `RedrawEventsCleared`, so somtimes it happens that you're finished them to early, and compositor hasn't send you frame event, but you won't get any new event, however once compositor will send you frame event, winit will wakeup and send you `RedrawEventsCleared`, because winit reading in a blocking way and has some kind of a `main` queue. if we dispatched something on this queue it means that we got frame event,  but since we have a situation, when you've already processes all events before frame event if you just check pending events you'll skip the redraw. The easiest way to repro it is moving with high repeat rate, since sometimes the last redraw is being missed.

Reply:
I'm not sure about all the points raised in your comments, but I think this basically boils down to `terminal.dirty` being `true`, while the event queue is empty.

So basically what we'd want to check is `event_queue_empty && (!redraw_requested || !terminal.dirty)`.

However, the problem with that is to check if the terminal is dirty we require locking it. If we lock it right here we need to drop it immediately and then re-lock it later just to render.

This means that in the worst case for the current solution, we'd lock it once and do all the following things for no reason:
 * create new processor
 * loop over all 0 events and process them
 * verify that there is indeed no display update pending

The worst case for changing the dirtiness would be that the terminal is locked twice with no other disadvantages. However the terminal lock is **very** expensive when there's currently a high PTY usage, however in that case there would be plenty of events lined up.

I feel like it might be worth it to lock the terminal, to save us a bunch of extra work. However I haven't benchmarked it.